n4js-libs/packages/n4js-cli/src/n4js/index.n4js

```diff
@@ -73,118 +68,233 @@ export public function n4jsc(
+ */
+export public async function createN4JSProcess(args: N4jscArguments, spawnOptions: childProcess.SpawnOptions, log: {function(string)} = npmlog.info) : childProcess.ChildProcess {
+    const modifedArgs = await modifyArgs(args, log);
+    const java = getJavaCommand(log);
+    return childProcess.spawn(java, modifedArgs.toArray(), spawnOptions);
+}
+
+
+
+async function modifyArgs(args: N4jscArguments, log: {function(string)}): N4jscArguments {
+    // set java option -Xmx
+    const envJavaToolOptions = process.env.JAVA_TOOL_OPTIONS as string;
+    const match = envJavaToolOptions?.match(/-xmx(?<size>[\S]*)/);
+    const size = match?.groups?.size;
+    if (size) {
+        let vmXmxSet = false;
+        for (let i=0; i<args.vmArgs.length && !vmXmxSet; i++) {
+            const arg = args.vmArgs[i];
+            if (arg.startsWith(JAVA_VM_XMX)) {
+                args.vmArgs[i] = JAVA_VM_XMX + size;
+                vmXmxSet = true;
+            }
+        }
+        if (!vmXmxSet) {
+            args.vmArgs.push(JAVA_VM_XMX + size);
+        }
+    }
+
+    // replace n4jsc.jar (and download if not available)
+    const n4jscjarConfigProperty = findN4jscJarConfigProperty(args.dir, log);
+    if (n4jscjarConfigProperty) {
+        try {
+            const n4jscjarName = await ensureCustomN4jscJar(n4jscjarConfigProperty, log);
+            if (n4jscjarName) {
+                log("using n4jsc.jar: " + n4jscjarName);
+                args.n4jscjar = n4jscjarName;
+            }
+        } catch (e) {
+            log("Failed to use jar specified in .n4js-cli.config. Using default jar instead.");
+        }
+    }
 
-    const n4jscArgs = getN4jscArgs(goal, dir, options, log);
-    const vmArgs = getVmArgs(vmOptions, log);
-    const args = ["-jar", ...vmArgs, N4JSC_JAR, ...n4jscArgs];
+    // lsp goal forbids dir argument, but n4js-cli uses this argument to support n4js-cli configuration files
+    if (Goal.lsp == args.goal) {
+        args.dir = undefined;
+    }
 
-    const n4jscOptionsString = options? Object.entries(options).map((entry) => Array.from(entry).join(":")).join(" ") : '';
-    const vmOptionsString = vmOptions? Object.entries(vmOptions).map((entry) => Array.from(entry).join(":")).join(" ") : '';
-    log(`goal=${goal}; dir=${dir}; options=${n4jscOptionsString}; vmOptions=${vmOptionsString}`);
+    // set verbose flag
+    if (debugLog.enabled) {
+        args.n4jscOptions.push("--verbose");
+    }
 
     return args;
 }
 
-function getVmArgs(vmOptions: ~VmOptions+, log: {function(string)}): string[] {
-    const args: string[] = [];
 
-    if (!vmOptions) {
-        vmOptions = {};
+/**
+ * Returns either:
+ * - the value of the environment variable 'N4JSC_JAR',
+ * - the value of property 'n4jscjar' of the n4js-cli configuration file,
+ * - or null.
+ */
+export public function findN4jscJarConfigProperty(workingDir: string, log: {function(string)}) : string {
```

Comment:
The VSCode extension passes in the VSCode workspace root folder as "workingDir". But usually the user has some flexibility of choosing the VSCode workspace root folder and with certain combinations a n4js-cli config file might be overlooked.

For example: n4js-cli config file is located in root folder of the yarn workspace project (= root folder of Git repo) and the user opens its parent folder as VSCode workspace root folder.

(if people put the n4js-cli config file into projects in the "packages" folder there are more such cases)